dbms/src/Storages/DeltaMerge/File/ColumnStream.cpp

```diff
@@ -0,0 +1,317 @@
+        reader.dmfile->configuration->getChecksumAlgorithm(),
+        reader.dmfile->configuration->getChecksumFrameLength());
+}
+
+std::unique_ptr<CompressedSeekableReaderBuffer> ColumnReadStream::buildColDataReadBuffByMetaV2(
+    DMFileReader & reader,
+    [[maybe_unused]] ColId col_id,
+    const String & file_name_base,
+    const ReadLimiterPtr & read_limiter)
+{
+    auto info = reader.dmfile->merged_sub_file_infos.find(reader.dmfile->colDataFileName(file_name_base));
+    if (info == reader.dmfile->merged_sub_file_infos.end())
+    {
+        return std::make_unique<CompressedReadBufferFromFileProvider<false>>(
+            reader.file_provider,
+            reader.dmfile->colDataPath(file_name_base),
+            reader.dmfile->encryptionDataPath(file_name_base),
+            reader.dmfile->getConfiguration()->getChecksumFrameLength(),
+            read_limiter,
+            reader.dmfile->configuration->getChecksumAlgorithm(),
+            reader.dmfile->configuration->getChecksumFrameLength());
+    }
+
+    assert(info != reader.dmfile->merged_sub_file_infos.end());
+    auto file_path = reader.dmfile->mergedPath(info->second.number);
+    auto encryp_path = reader.dmfile->encryptionMergedPath(info->second.number);
+    auto offset = info->second.offset;
+    auto size = info->second.size;
+
+    // first read from merged file to get the raw data(contains the header)
+    auto buffer = ReadBufferFromRandomAccessFileBuilder::build(
+        reader.file_provider,
+        file_path,
+        encryp_path,
+        reader.dmfile->getConfiguration()->getChecksumFrameLength(),
+        read_limiter);
+    buffer.seek(offset);
+
+    String raw_data;
+    raw_data.resize(size);
+
+    buffer.read(reinterpret_cast<char *>(raw_data.data()), size);
+    // read from the buffer based on the raw data
+    return std::make_unique<CompressedReadBufferFromFileProvider</*has_checksum=*/false>>(
+        std::move(raw_data),
+        file_path,
+        reader.dmfile->getConfiguration()->getChecksumFrameLength(),
+        reader.dmfile->configuration->getChecksumAlgorithm(),
+        reader.dmfile->configuration->getChecksumFrameLength());
+}
+
+ColumnReadStream::ColumnReadStream(
+    DMFileReader & reader,
+    ColId col_id,
+    const String & file_name_base,
+    size_t max_read_buffer_size,
+    const LoggerPtr & log,
+    const ReadLimiterPtr & read_limiter)
+    : avg_size_hint(reader.dmfile->getColumnStat(col_id).avg_size)
+{
+    // load mark data
+    // TODO: Do we still need `marks` if the reader don't contains any pack?
+    if (reader.mark_cache)
+        marks = reader.mark_cache->getOrSet(
+            reader.dmfile->colMarkCacheKey(file_name_base),
+            MarkLoader(reader, col_id, file_name_base, read_limiter));
+    else
+        marks = MarkLoader(reader, col_id, file_name_base, read_limiter)();
+
+    // skip empty dmfile
+    size_t packs = reader.dmfile->getPacks();
+    if (packs == 0)
+        return;
```

Comment:
we can check `pack_filter` here, if `pack_filter->getUsePacks()` contains all 0, we can skip create read buffer?

Reply:
I think we can do it later for this issue. https://github.com/pingcap/tiflash/issues/8759
This PR focuses on refactoring the codes.

Reply:
ok